src/databricks/labs/ucx/progress/history.py

```diff
@@ -0,0 +1,216 @@
+from __future__ import annotations
+import dataclasses
+import datetime as dt
+import json
+import logging
+from collections.abc import Iterable, Sequence
+from typing import ClassVar, Protocol, TypeVar, Generic, Any, get_type_hints
+
+from databricks.labs.lsql.backends import SqlBackend
+
+from databricks.labs.ucx.framework.owners import Ownership
+from databricks.labs.ucx.framework.utils import escape_sql_identifier
+from databricks.labs.ucx.progress.install import Historical
+
+logger = logging.getLogger(__name__)
+
+
+class DataclassWithIdAttributes(Protocol):
+    __dataclass_fields__: ClassVar[dict[str, Any]]
+
+    __id_attributes__: ClassVar[tuple[str, ...]]
+    """The names of attributes (can be dataclass fields or ordinary properties) that make up the object identifier.
+
+    All attributes must be (non-optional) strings.
+    """
+
+
+Record = TypeVar("Record", bound=DataclassWithIdAttributes)
+
+
+class HistoricalEncoder(Generic[Record]):
```

Comment:
Could you add a docstring on what this class does and why it is important/required?
 
 Important because: we cannot simply convert records to historical data by `json.dumps(dataclasses.asdict(record)))` because we:
 - Need to handle `failures` separately
 - ...

Reply:
Excellent suggestion. I've added this, does it seem right to you?

Reply:
It looks good! I have some pointers, but general looks good